src/opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.cpp

```diff
@@ -67,176 +56,232 @@ double GuideRate::RateVector::eval(GuideRateModel::Target target) const {
     if (target == GuideRateModel::Target::WAT)
         return this->wat_rat;
 
-    throw std::logic_error("Don't know how to convert .... ");
+    throw std::logic_error {
+        "Don't know how to convert target type " + std::to_string(static_cast<int>(target))
+    };
 }
 
 
 GuideRate::GuideRate(const Schedule& schedule_arg) :
     schedule(schedule_arg)
 {}
 
-
-
-double GuideRate::get(const std::string& well, Well::GuideRateTarget target, const RateVector& rates) const {
-    auto model_target = GuideRateModel::convert_target(target);
-    return get(well, model_target, rates);
+double GuideRate::get(const std::string& well, Well::GuideRateTarget target, const RateVector& rates) const
+{
+    return this->get(well, GuideRateModel::convert_target(target), rates);
 }
 
-double GuideRate::get(const std::string& group, Group::GuideRateTarget target, const RateVector& rates) const {
-    auto model_target = GuideRateModel::convert_target(target);
-    return get(group, model_target, rates);
+double GuideRate::get(const std::string& group, Group::GuideRateTarget target, const RateVector& rates) const
+{
+    return this->get(group, GuideRateModel::convert_target(target), rates);
 }
 
-double GuideRate::get(const std::string& name, GuideRateModel::Target model_target, const RateVector& rates) const {
-    const auto iter = this->values.find(name);
-    if (iter != this->values.end()) {
-        const auto& value = iter->second;
-        if (value.target == model_target)
-            return value.value;
-        else {
-            double model_target_rate = rates.eval(model_target);
-            double value_target_rate = rates.eval(value.target);        
-                    
-            if (model_target_rate < 1e-6)
-                return value.value;
-            if (value_target_rate < 1e-6)
-                return value.value; 
+double GuideRate::get(const std::string& name, GuideRateModel::Target model_target, const RateVector& rates) const
+{
+    using namespace unit;
+    using prefix::micro;
 
-            // scale with the current production ratio when the control target differs from the guide rate target.
-            return value.value * model_target_rate / value_target_rate;
-        }
-    } else {
-        const auto& pot = this->potentials.at(name);
-        return pot.eval(model_target);
+    auto iter = this->values.find(name);
+    if (iter == this->values.end()) {
+        return this->potentials.at(name).eval(model_target);
+    }
+
+    const auto& value = *iter->second;
+    const auto grvalue = this->get_grvalue_result(value);
+    if (value.curr.target == model_target) {
+        return grvalue;
+    }
+
+    const auto value_target_rate = rates.eval(value.curr.target);
+    if (value_target_rate < 1.0*micro*liter/day) {
```

Comment:
What is this limit?

Reply:
> What is this limit?

It is admittedly a hack.

To be fair, the condition avoids division by zero in the [`scale`](https://github.com/OPM/opm-common/blob/3efae325003fa7b619ede8781972ff47eb293de1/src/opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.cpp#L102) variable, I'd **really** like to have something better here.  On the other hand this check is solely responsible for the nonsense spikes in the water guiderates just as the well comes online.  I tried removing it completely, but that ended up having deleterious downstream effects and threw the real total production measure (e.g., `FOPR` or `GOPR`) completely off from the reference case.

The current master sources use a hard-coded value of [`1.0e-6`](https://github.com/OPM/opm-common/blob/20a6a522dc540f99ebfc7bca5b68311169e4cd82/src/opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.cpp#L100-L103) here and since we're compare SI rates at this point that felt too much like a threshold pulled out of thin air.  Granted 1 microlitre per day is also a number pulled out of thin air, but it does at least suggest a small rate value in all unit systems.